gazebo_ros2_control/src/default_robot_hw_sim.cpp

```diff
@@ -1,624 +0,0 @@
-      RCLCPP_WARN_STREAM(nh_->get_logger(), "cant register opmode handle for joint" << joint_name);
-    }
-  }
-
-  // since handles references may have changed due to underlying DynamicJointState msg
-  // vectors resizing and reallocating we need to get these handles again
-  // any handles not registered are skipped, such as the command handles if they arent
-  // involved in the control method
-  bindJointHandles(joint_pos_stateh_);
-  bindJointHandles(joint_vel_stateh_);
-  bindJointHandles(joint_eff_stateh_);
-  bindJointHandles(joint_pos_cmdh_);
-  bindJointHandles(joint_vel_cmdh_);
-  bindJointHandles(joint_eff_cmdh_);
-
-
-  //
-  // Complete initialization of limits, PID controllers, etc now that registered handles are bound
-  //
-  for (size_t j = 0; j < joint_names_.size(); j++) {
-    auto simjoint = sim_joints_[j];
-    assert(simjoint);
-
-    // setup joint limits
-    registerJointLimits(
-      j,
-      urdf_model,
-      &joint_types_[j], &joint_lower_limits_[j], &joint_upper_limits_[j],
-      &joint_effort_limits_[j], &joint_vel_limits_[j]);
-    if (joint_control_methods_[j] != EFFORT) {
-      try {
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".p");
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".i");
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".d");
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".i_clamp_max");
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".i_clamp_min");
-        nh_->declare_parameter(transmissions[j].joints[0].name + ".antiwindup", false);
-
-        if (nh_->get_parameter(transmissions[j].joints[0].name + ".p").get_type() ==
-          rclcpp::PARAMETER_DOUBLE &&
-          nh_->get_parameter(transmissions[j].joints[0].name + ".i").get_type() ==
-          rclcpp::PARAMETER_DOUBLE &&
-          nh_->get_parameter(transmissions[j].joints[0].name + ".d").get_type() ==
-          rclcpp::PARAMETER_DOUBLE)
-        {
-          pid_controllers_.push_back(
-            control_toolbox::PidROS(nh_, transmissions[j].joints[0].name));
-          if (pid_controllers_[j].initPid()) {
-            switch (joint_control_methods_[j]) {
-              case POSITION: joint_control_methods_[j] = POSITION_PID;
-                RCLCPP_INFO(
-                  nh_->get_logger(), "joint %s is configured in POSITION_PID mode",
-                  transmissions[j].joints[0].name.c_str());
-                break;
-              case VELOCITY: joint_control_methods_[j] = VELOCITY_PID;
-                RCLCPP_INFO(
-                  nh_->get_logger(), "joint %s is configured in VELOCITY_PID mode",
-                  transmissions[j].joints[0].name.c_str());
-                break;
-            }
-          }
-        }
-      } catch (const std::exception & e) {
-        RCLCPP_ERROR(rclcpp::get_logger("gazebo_ros2_control"), "%s", e.what());
-      }
-      simjoint->SetParam("fmax", 0, joint_effort_limits_[j]);
```

Comment:
@ahcorde This is making the difference for me. I comment out this line, then the main branch code is having the same strange behavior as this PR.

Does gazebo not respect the `effort` set inside `<limit effort=""/>` from the URDF? We still need this for position control.

Reply:
Right now there is no `joint_limit_interface` available in `ros2_control`. There is a PR openned https://github.com/ros-controls/ros2_control/pull/271 to include this again in the `hardware_interace`.

Reply:
Thanks. So the problem will not be there after `joint_limit_interface` is implemented. I previously didn't understand what was causing the behaviour that I was facing. You can resolve this thread now.